File: frontend/src/lib/math/affordability.ts
import { pmt, fv } from './finance';

// ── FOIR table (pp2 #3 — single source of truth) ────────────────────────────
// All verdict and lever logic must call foirBandFor(); no inline cap math elsewhere.

export type EmploymentType = 'salaried' | 'self_employed';
export type LoanType = 'home' | 'personal' | 'vehicle' | 'other';

const FOIR_CAPS: Record<EmploymentType, Record<LoanType, number>> = {
  salaried: { home: 0.5, personal: 0.45, vehicle: 0.45, other: 0.4 },
  self_employed: { home: 0.45, personal: 0.4, vehicle: 0.4, other: 0.4 },
};

export function foirBandFor(
  netMonthly: number,
  employmentType: EmploymentType = 'salaried',
  loanType: LoanType = 'other',
): { cap: number; capAmount: number; label: string } {
  const cap = FOIR_CAPS[employmentType][loanType];
  return {
    cap,
    capAmount: Math.round(netMonthly * cap),
    label: `${Math.round(cap * 100)}% of net income`,
  };
}

// ── Types ────────────────────────────────────────────────────────────────────

export type LeverType = 'increaseSurplus' | 'cutExpenses' | 'extendTenure' | 'raiseIncome';

export type BlockedReason = 'foir_cap' | 'retirement_cap' | 'surplus_floor';

export interface Lever {
  type: LeverType;
  /** Additional monthly savings needed to reach target within benchmark window */
  monthlySavingsNeeded?: number;
  /** Months to afford after applying this lever */
  newMonthsToAfford?: number;
  /** Income level needed for this lever to close the gap */
  targetIncome?: number;
  /** Extended tenure (EMI route) */
  newTenureMonths?: number;
  /** Why the current path is blocked */
  reasonIfBlocked?: BlockedReason;
}

export interface AffordabilityInput {
  targetCost: number;
  route: 'cash' | 'emi';
  netMonthlyIncome: number;
  /** ExpenseProfile.total — should NOT include debt EMIs */
  monthlyExpenses: number;
  /** monthlySurplusReserve — intentional monthly park */
  monthlyReserve: number;
  /** DebtProfile.totalMonthly — existing EMI obligations */
  existingEmiTotal: number;
  /** Annual % return on savings (e.g. 12). Used for cash-route FV. */
  investmentReturnRate: number;
  // EMI route
  /** Annual interest rate % (default 9) */
  annualInterestRate?: number;
  /** Loan tenure in months (default 60) */
  tenureMonths?: number;
  // Optional Tier-2 signals
  ageYears?: number;
  employmentType?: EmploymentType;
  loanType?: LoanType;
}

export interface AffordabilityResult {
  verdict: 'affordable_now' | 'affordable_later' | 'not_affordable';
  /** Cash route: months of saving needed. null if EMI route or not computable. */
  monthsToAfford: number | null;
  monthlySurplus: number;
  /** EMI route: computed monthly payment via PMT. */
  emi: number | null;
  /** EMI route: whether total EMIs fit within FOIR cap. null for cash route. */
  foirOk: boolean | null;
  /** EMI route: the FOIR cap amount (₹) for display. null for cash route. */
  foirCapAmount: number | null;
  /** Monthly gap to close: 0 if affordable_now, else shortfall. */
  gap: number;
  levers: Lever[];
}

// ── Helpers ──────────────────────────────────────────────────────────────────

const BENCHMARK_MONTHS = 36; // "affordable in 3 years" target for lever math

/**
 * Analytic months-to-afford for cash savings.
 * Returns null when surplus ≤ 0 (can't save at all).
 */
function monthsToSave(
  targetCost: number,
  monthlySurplus: number,
  monthlyRate: number,
): number | null {
  if (monthlySurplus <= 0) return null;
  if (monthlyRate === 0) return Math.ceil(targetCost / monthlySurplus);
  // Solve FV = targetCost for n: n = log(1 + targetCost*r/surplus) / log(1+r)
  const n = Math.log(1 + (targetCost * monthlyRate) / monthlySurplus) / Math.log(1 + monthlyRate);
  return Math.ceil(n);
}

/**
 * Monthly savings needed to accumulate targetCost in exactly n months.
 * Inverse of FV formula.
 */
function requiredMonthlySavings(targetCost: number, months: number, monthlyRate: number): number {
  if (months <= 0) return targetCost;
  if (monthlyRate === 0) return Math.ceil(targetCost / months);
  const growth = Math.pow(1 + monthlyRate, months);
  return Math.ceil((targetCost * monthlyRate) / (growth - 1));
}

function buildCashLevers(
  input: AffordabilityInput,
  surplus: number,
  monthsToAfford: number | null,
  monthlyRate: number,
): Lever[] {
  const levers: Lever[] = [];

  // No levers needed when affordable within benchmark window
  if (monthsToAfford !== null && monthsToAfford <= BENCHMARK_MONTHS) return levers;

  const required36 = requiredMonthlySavings(input.targetCost, BENCHMARK_MONTHS, monthlyRate);
  const gap36 = required36 - Math.max(0, surplus);

  if (gap36 > 0) {
    // increaseSurplus — save more each month
    levers.push({
      type: 'increaseSurplus',
      monthlySavingsNeeded: gap36,
      newMonthsToAfford: BENCHMARK_MONTHS,
    });

    // cutExpenses — same math, different framing
    levers.push({
      type: 'cutExpenses',
      monthlySavingsNeeded: gap36,
      newMonthsToAfford: BENCHMARK_MONTHS,
    });

    // raiseIncome — what income level achieves the benchmark-month surplus
    const targetIncome =
      required36 + input.monthlyExpenses + input.monthlyReserve + input.existingEmiTotal;
    if (targetIncome > input.netMonthlyIncome) {
      levers.push({
        type: 'raiseIncome',
        targetIncome,
        newMonthsToAfford: BENCHMARK_MONTHS,
        reasonIfBlocked: surplus <= 0 ? 'surplus_floor' : undefined,
      });
    }
  }

  return levers;
}

function buildEmiLevers(
  input: AffordabilityInput,
  surplus: number,
  emi: number,
  foirOk: boolean,
  foirCapAmount: number,
  effectiveTenure: number,
  monthlyRate: number,
): Lever[] {
  const levers: Lever[] = [];
  const loanType = input.loanType ?? 'other';
  const employmentType = input.employmentType ?? 'salaried';

  if (!foirOk) {
    // extendTenure — find min tenure where EMI fits FOIR cap
    const maxTenure = input.ageYears ? (60 - input.ageYears) * 12 : 360;
    let extendedTenure: number | null = null;
    for (let t = effectiveTenure + 12; t <= maxTenure; t += 12) {
      const extEmi = pmt(monthlyRate, t, input.targetCost);
      const { capAmount } = foirBandFor(input.netMonthlyIncome, employmentType, loanType);
      if (input.existingEmiTotal + extEmi <= capAmount) {
        extendedTenure = t;
        break;
      }
    }

    if (extendedTenure !== null) {
      levers.push({
        type: 'extendTenure',
        newTenureMonths: extendedTenure,
        reasonIfBlocked:
          input.ageYears && extendedTenure >= (60 - input.ageYears) * 12
            ? 'retirement_cap'
            : 'foir_cap',
      });
    }

    // raiseIncome — income needed so existing + new EMI fits FOIR
    const incomeNeeded = Math.ceil(
      (input.existingEmiTotal + emi) / FOIR_CAPS[employmentType][loanType],
    );
    if (incomeNeeded > input.netMonthlyIncome) {
      levers.push({
        type: 'raiseIncome',
        targetIncome: incomeNeeded,
        reasonIfBlocked: 'foir_cap',
      });
    }
  } else if (emi > surplus) {
    // FOIR ok but EMI exceeds current surplus — need to free up cash flow
    const surplusGap = emi - surplus;

    levers.push({
      type: 'cutExpenses',
      monthlySavingsNeeded: surplusGap,
    });

    levers.push({
      type: 'raiseIncome',
      targetIncome: input.netMonthlyIncome + surplusGap,
    });
  }

  return levers;
}

// ── Main engine ──────────────────────────────────────────────────────────────

export function runAffordability(input: AffordabilityInput): AffordabilityResult {
  const {
    targetCost,
    route,
    netMonthlyIncome,
    monthlyExpenses,
    monthlyReserve,
    existingEmiTotal,
    investmentReturnRate,
    annualInterestRate = 9,
    tenureMonths = 60,
    ageYears,
    employmentType = 'salaried',
    loanType = 'other',
  } = input;

  // monthlySurplus: what's left after expenses, existing EMIs, and reserve
  const monthlySurplus = netMonthlyIncome - monthlyExpenses - existingEmiTotal - monthlyReserve;
  const monthlyRate = investmentReturnRate / 12 / 100;

  if (route === 'cash') {
    const mta = monthsToSave(targetCost, monthlySurplus, monthlyRate);
    const verdict =
      mta !== null && mta <= 1
        ? 'affordable_now'
        : mta !== null
          ? 'affordable_later'
          : 'not_affordable';

    const gap =
      verdict === 'affordable_now'
        ? 0
        : monthlySurplus <= 0
          ? Math.abs(monthlySurplus) +
            requiredMonthlySavings(targetCost, BENCHMARK_MONTHS, monthlyRate)
          : Math.max(
              0,
              requiredMonthlySavings(targetCost, BENCHMARK_MONTHS, monthlyRate) - monthlySurplus,
            );

    const levers = buildCashLevers(input, monthlySurplus, mta, monthlyRate);

    return {
      verdict,
      monthsToAfford: mta,
      monthlySurplus,
      emi: null,
      foirOk: null,
      foirCapAmount: null,
      gap,
      levers,
    };
  }

  // EMI route
  // Cap tenure at retirement if age provided
  const maxTenure = ageYears ? Math.max(1, (60 - ageYears) * 12) : tenureMonths;
  const effectiveTenure = Math.min(tenureMonths, maxTenure);
  const emiRate = annualInterestRate / 12 / 100;
  const emi = pmt(emiRate, effectiveTenure, targetCost);

  const { capAmount: foirCapAmount } = foirBandFor(netMonthlyIncome, employmentType, loanType);
  const foirOk = input.existingEmiTotal + emi <= foirCapAmount;

  const gap = foirOk
    ? Math.max(0, emi - monthlySurplus)
    : Math.max(0, input.existingEmiTotal + emi - foirCapAmount);

  const verdict =
    foirOk && emi <= monthlySurplus
      ? 'affordable_now'
      : foirOk || gap > 0
        ? 'affordable_later'
        : 'not_affordable';

  const levers = buildEmiLevers(
    input,
    monthlySurplus,
    emi,
    foirOk,
    foirCapAmount,
    effectiveTenure,
    emiRate,
  );

  // not_affordable only when FOIR blocked AND no lever can fix it
  const finalVerdict = !foirOk && levers.length === 0 ? 'not_affordable' : verdict;

  return {
    verdict: finalVerdict,
    monthsToAfford: null,
    monthlySurplus,
    emi,
    foirOk,
    foirCapAmount,
    gap,
    levers,
  };
}

File: backend/app/engines/affordability.py
"""Python port of frontend/src/lib/math/affordability.ts.

Parity notes:
- pmt/fv use js_round (Math.round semantics) to match TS outputs.
- monthsToSave / requiredMonthlySavings use math.ceil (same as JS Math.ceil).
- FOIR caps and BENCHMARK_MONTHS are identical to the TS constants.
"""

from __future__ import annotations

import math
from typing import Any, Literal

from app.engines._helpers import js_round

# ── FOIR table ───────────────────────────────────────────────────────────────

EmploymentType = Literal["salaried", "self_employed"]
LoanType = Literal["home", "personal", "vehicle", "other"]

_FOIR_CAPS: dict[str, dict[str, float]] = {
    "salaried": {"home": 0.5, "personal": 0.45, "vehicle": 0.45, "other": 0.4},
    "self_employed": {"home": 0.45, "personal": 0.4, "vehicle": 0.4, "other": 0.4},
}

BENCHMARK_MONTHS = 36


def foir_band_for(
    net_monthly: float,
    employment_type: str = "salaried",
    loan_type: str = "other",
) -> dict[str, Any]:
    cap = _FOIR_CAPS.get(employment_type, _FOIR_CAPS["salaried"]).get(
        loan_type, _FOIR_CAPS["salaried"]["other"]
    )
    return {
        "cap": cap,
        "capAmount": js_round(net_monthly * cap),
        "label": f"{round(cap * 100)}% of net income",
    }


# ── Financial math helpers ───────────────────────────────────────────────────


def _pmt(rate: float, nper: int, pv: float) -> int:
    """Monthly payment on a loan. Mirrors TS pmt()."""
    if nper <= 0:
        return 0
    if rate == 0:
        return js_round(pv / nper)
    payment = (pv * rate) / (1 - (1 + rate) ** -nper)
    return js_round(payment)


def _months_to_save(
    target_cost: float,
    monthly_surplus: float,
    monthly_rate: float,
) -> int | None:
    """Analytic months-to-afford for cash savings. Returns None when surplus ≤ 0."""
    if monthly_surplus <= 0:
        return None
    if monthly_rate == 0:
        return math.ceil(target_cost / monthly_surplus)
    n = math.log(1 + (target_cost * monthly_rate) / monthly_surplus) / math.log(1 + monthly_rate)
    return math.ceil(n)


def _required_monthly_savings(
    target_cost: float,
    months: int,
    monthly_rate: float,
) -> int:
    """Monthly savings needed to accumulate target_cost in exactly n months."""
    if months <= 0:
        return int(target_cost)
    if monthly_rate == 0:
        return math.ceil(target_cost / months)
    growth = (1 + monthly_rate) ** months
    return math.ceil((target_cost * monthly_rate) / (growth - 1))


# ── Lever builders ───────────────────────────────────────────────────────────


def _build_cash_levers(
    inp: dict[str, Any],
    surplus: float,
    months_to_afford: int | None,
    monthly_rate: float,
) -> list[dict[str, Any]]:
    if months_to_afford is not None and months_to_afford <= BENCHMARK_MONTHS:
        return []

    target_cost: float = inp["targetCost"]
    levers: list[dict[str, Any]] = []
    required36 = _required_monthly_savings(target_cost, BENCHMARK_MONTHS, monthly_rate)
    gap36 = required36 - max(0.0, surplus)

    if gap36 > 0:
        levers.append(
            {
                "type": "increaseSurplus",
                "monthlySavingsNeeded": int(gap36),
                "newMonthsToAfford": BENCHMARK_MONTHS,
            }
        )
        levers.append(
            {
                "type": "cutExpenses",
                "monthlySavingsNeeded": int(gap36),
                "newMonthsToAfford": BENCHMARK_MONTHS,
            }
        )
        target_income = (
            required36 + inp["monthlyExpenses"] + inp["monthlyReserve"] + inp["existingEmiTotal"]
        )
        if target_income > inp["netMonthlyIncome"]:
            lever: dict[str, Any] = {
                "type": "raiseIncome",
                "targetIncome": math.ceil(target_income),
                "newMonthsToAfford": BENCHMARK_MONTHS,
            }
            if surplus <= 0:
                lever["reasonIfBlocked"] = "surplus_floor"
            levers.append(lever)

    return levers


def _build_emi_levers(
    inp: dict[str, Any],
    surplus: float,
    emi: int,
    foir_ok: bool,
    foir_cap_amount: int,
    effective_tenure: int,
    monthly_rate: float,
) -> list[dict[str, Any]]:
    levers: list[dict[str, Any]] = []
    loan_type = inp.get("loanType") or "other"
    employment_type = inp.get("employmentType") or "salaried"
    age_years: int | None = inp.get("ageYears")

    if not foir_ok:
        max_tenure = (60 - age_years) * 12 if age_years else 360
        extended_tenure: int | None = None
        t = effective_tenure + 12
        while t <= max_tenure:
            ext_emi = _pmt(monthly_rate, t, inp["targetCost"])
            band = foir_band_for(inp["netMonthlyIncome"], employment_type, loan_type)
            if inp["existingEmiTotal"] + ext_emi <= band["capAmount"]:
                extended_tenure = t
                break
            t += 12

        if extended_tenure is not None:
            reason: str = "foir_cap"
            if age_years and extended_tenure >= (60 - age_years) * 12:
                reason = "retirement_cap"
            levers.append(
                {
                    "type": "extendTenure",
                    "newTenureMonths": extended_tenure,
                    "reasonIfBlocked": reason,
                }
            )

        emp_foir_cap = _FOIR_CAPS.get(employment_type, _FOIR_CAPS["salaried"]).get(
            loan_type, _FOIR_CAPS["salaried"]["other"]
        )
        income_needed = math.ceil((inp["existingEmiTotal"] + emi) / emp_foir_cap)
        if income_needed > inp["netMonthlyIncome"]:
            levers.append(
                {
                    "type": "raiseIncome",
                    "targetIncome": income_needed,
                    "reasonIfBlocked": "foir_cap",
                }
            )

    elif emi > surplus:
        surplus_gap = emi - surplus
        levers.append({"type": "cutExpenses", "monthlySavingsNeeded": math.ceil(surplus_gap)})
        levers.append(
            {
                "type": "raiseIncome",
                "targetIncome": math.ceil(inp["netMonthlyIncome"] + surplus_gap),
            }
        )

    return levers


# ── Main engine ──────────────────────────────────────────────────────────────


def run_affordability(inp: dict[str, Any]) -> dict[str, Any]:
    """Port of runAffordability() in frontend/src/lib/math/affordability.ts.

    Input keys (camelCase, matching TS AffordabilityInput):
      targetCost, route, netMonthlyIncome, monthlyExpenses, monthlyReserve,
      existingEmiTotal, investmentReturnRate, annualInterestRate (default 9),
      tenureMonths (default 60), ageYears?, employmentType?, loanType?

    Returns dict matching AffordabilityResult shape.
    """
    target_cost: float = float(inp["targetCost"])
    route: str = inp.get("route", "cash")
    net_monthly_income: float = float(inp["netMonthlyIncome"])
    monthly_expenses: float = float(inp["monthlyExpenses"])
    monthly_reserve: float = float(inp.get("monthlyReserve", 0))
    existing_emi_total: float = float(inp.get("existingEmiTotal", 0))
    investment_return_rate: float = float(inp.get("investmentReturnRate", 8))
    annual_interest_rate: float = float(inp.get("annualInterestRate", 9))
    tenure_months: int = int(inp.get("tenureMonths", 60))
    age_years: int | None = inp.get("ageYears")
    employment_type: str = inp.get("employmentType") or "salaried"
    loan_type: str = inp.get("loanType") or "other"

    monthly_surplus = net_monthly_income - monthly_expenses - existing_emi_total - monthly_reserve
    monthly_rate = investment_return_rate / 12 / 100

    if route == "cash":
        mta = _months_to_save(target_cost, monthly_surplus, monthly_rate)
        if mta is not None and mta <= 1:
            verdict = "affordable_now"
        elif mta is not None:
            verdict = "affordable_later"
        else:
            verdict = "not_affordable"

        if verdict == "affordable_now":
            gap = 0
        elif monthly_surplus <= 0:
            gap = abs(monthly_surplus) + _required_monthly_savings(
                target_cost, BENCHMARK_MONTHS, monthly_rate
            )
        else:
            gap = max(
                0,
                _required_monthly_savings(target_cost, BENCHMARK_MONTHS, monthly_rate)
                - monthly_surplus,
            )

        levers = _build_cash_levers(inp, monthly_surplus, mta, monthly_rate)

        return {
            "verdict": verdict,
            "monthsToAfford": mta,
            "monthlySurplus": monthly_surplus,
            "emi": None,
            "foirOk": None,
            "foirCapAmount": None,
            "gap": gap,
            "levers": levers,
        }

    # EMI route
    if age_years:
        max_tenure = max(1, (60 - age_years) * 12)
    else:
        max_tenure = tenure_months
    effective_tenure = min(tenure_months, max_tenure)
    emi_rate = annual_interest_rate / 12 / 100
    emi = _pmt(emi_rate, effective_tenure, target_cost)

    band = foir_band_for(net_monthly_income, employment_type, loan_type)
    foir_cap_amount: int = band["capAmount"]
    foir_ok = (existing_emi_total + emi) <= foir_cap_amount

    if foir_ok:
        gap = max(0, emi - monthly_surplus)
    else:
        gap = max(0, existing_emi_total + emi - foir_cap_amount)

    if foir_ok and emi <= monthly_surplus:
        verdict = "affordable_now"
    elif foir_ok or gap > 0:
        verdict = "affordable_later"
    else:
        verdict = "not_affordable"

    levers = _build_emi_levers(
        inp, monthly_surplus, emi, foir_ok, foir_cap_amount, effective_tenure, emi_rate
    )

    # not_affordable only when FOIR blocked AND no lever can fix it
    final_verdict = "not_affordable" if not foir_ok and len(levers) == 0 else verdict

    return {
        "verdict": final_verdict,
        "monthsToAfford": None,
        "monthlySurplus": monthly_surplus,
        "emi": emi,
        "foirOk": foir_ok,
        "foirCapAmount": foir_cap_amount,
        "gap": gap,
        "levers": levers,
    }
